### ISSUE TITLE
Is this compatible with TS 5 decorators?

### ISSUE THREAD

**Issue by ffMathy** [microsoft/tsyringe#225]
Typescript 5.0 recently released. It changes decorators quite a bit.

Will this library work without the experimental decorators flag? I think that should be written in the Readme somewhere.

**Comment by JorgeLNJunior** [comment]
According to [Typescript 5 announcement](https://devblogs.microsoft.com/typescript/announcing-typescript-5-0-beta/#differences-with-experimental-legacy-decorators) new decorators are not compatible with `emitDecoratorMetadata`. So no, the experimental decorators flag is still required.

**Comment by South-Paw** [comment]
Is there any plans by the team to update this package to support TS 5.0?

**Comment by rfire12** [comment]
None of the maintainers seem to be working at microsoft. Not sure if they ever did, as I just started using this library recently. 

So I guess the future of this project seem to be uncertain, as there hasn't been much activity either lately.

**Comment by South-Paw** [comment]
> None of the maintainers seem to be working at microsoft. Not sure if they ever did, as I just started using this library recently.
> 
> So I guess the future of this project seem to be uncertain, as there hasn't been much activity either lately.

The library is stable, has all the features it needs implemented and has been in that state for ages.

**Comment by stevenulibarri** [comment]
> > None of the maintainers seem to be working at microsoft. Not sure if they ever did, as I just started using this library recently.
> > So I guess the future of this project seem to be uncertain, as there hasn't been much activity either lately.
> 
> The library is stable, has all the features it needs implemented and has been in that state for ages.

Technically yes, but I think we should expect TS to cease supporting experimental decorators eventually. But its more likely that users of tsyringe may eventually encounter a package they want to use that requires the new decorators but can't until it becomes compatible.
But even then, that compatibility is not trivial, the new decorator spec does not support parameter injection at all and probably won't for a long time. In my own case its not a ton of work to remove instances of parameter injection from my projects but for others it might be more difficult.
Its an unfortunate situation all around.

**Comment by wenindoubt** [comment]
Does anyone have experience using another stable, lightweight DI framework that is more actively maintained?

**Comment by ffMathy** [comment]
I ended up going for Inversify.

**Comment by South-Paw** [comment]
> > > None of the maintainers seem to be working at microsoft. Not sure if they ever did, as I just started using this library recently.
> > > So I guess the future of this project seem to be uncertain, as there hasn't been much activity either lately.
> > 
> > 
> > The library is stable, has all the features it needs implemented and has been in that state for ages.
> 
> Technically yes, but I think we should expect TS to cease supporting experimental decorators eventually. But its more likely that users of tsyringe may eventually encounter a package they want to use that requires the new decorators but can't until it becomes compatible. But even then, that compatibility is not trivial, the new decorator spec does not support parameter injection at all and probably won't for a long time. In my own case its not a ton of work to remove instances of parameter injection from my projects but for others it might be more difficult. Its an unfortunate situation all around.

I agree - I want to see this package support TS 5 decorators as well or a fork that does and works with TS 5 so we aren't locked out of other newer decorator packages.

However let's not forget that we were all using a flag that does actually say it is [`experimental` on it's doc page](https://www.typescriptlang.org/tsconfig#emitDecoratorMetadata) and as such we knew this could or would happen eventually when decorator support got closer to being finalised 😄

On my reply, it was only to the comment of "future of this project seem to be uncertain, as there hasn't been much activity either lately". Just because a project doesn't have recent activity, doesn't mean its broken or out of date. Some libraries you have in your dependency trees haven't had updates for years and remain stable, secure and reliable despite this fact.

But let's try avoid getting into that whole topic, instead let's just focus on what we do with this package and TS 5 👍

**Comment by jhechtf** [comment]
In looking through the code it seems like this may have already been resolved via https://github.com/microsoft/tsyringe/pull/226 but it simply hasn't been published yet. 

Hopefully once that becomes published it should work without issue in TS 5.x environments.

**Comment by jhechtf** [comment]
Are there any plans to release the code changes that would publish the TS5 compatibility any time soon? PR looks like it was merged some time ago.

**Comment by PodaruDragos** [comment]
> I ended up going for Inversify.

inversify also uses Reflect.metadata so it won't work with TS 5 decorators.

**Comment by jhechtf** [comment]
> > I ended up going for Inversify.
> 
> inversify also uses Reflect.metadata so it won't work with TS 5 decorators.

Just tested, inversify is not affected by this with version 6.0.1

**Comment by nazrhyn** [comment]
@MeltingMosaic had merge capability, but sometimes it's different people who can publish new versions to NPM. I guess they'd also have to evaluate whether the other changes in `master` since the last published version are okay to publish.

**Comment by Aldiwildan77** [comment]
Any updates for these releases?

**Comment by mprimmaz** [comment]
> Any updates for these releases?

https://github.com/microsoft/tsyringe/pull/226

**Comment by yatiac** [comment]
Upgraded TS to v5 and now I am facing this 😢

**Comment by stevenulibarri** [comment]
I don't really understand what is happening with the repo. The PR that would allow moving to the latest TS version was merged and the action ran for the release, its green and everything, but NPM still shows 4.7.0 which is over a year old now.
https://github.com/microsoft/tsyringe/actions/runs/4716091295

**Comment by jhechtf** [comment]
> I don't really understand what is happening with the repo. The PR that would allow moving to the latest TS version was merged and the action ran for the release, its green and everything, but NPM still shows 4.7.0 which is over a year old now. https://github.com/microsoft/tsyringe/actions/runs/4716091295

In looking at that job you mentioned, the steps required to publish the package never ran, so the workflow didn't actually end up doing anything.

<img width="1372" alt="Screenshot 2023-05-11 at 11 48 52 AM" src="https://github.com/microsoft/tsyringe/assets/8152471/c9b96f78-c12a-4f20-ac08-1de33affb1e0">

This means that someone from the team needs to push a publishing commit so that these changes can go live.

**Comment by stevenulibarri** [comment]
I am unfamiliar with github actions so I wasn't sure if that meant they were disabled or if I just didn't have access to see more detail about the steps. In any case hopefully a maintainer can just get it shipped; I really don't want to have to fork anything or look for other options if it can be avoided.

**Comment by jhechtf** [comment]
> I am unfamiliar with github actions so I wasn't sure if that meant they were disabled or if I just didn't have access to see more detail about the steps. In any case hopefully a maintainer can just get it shipped; I really don't want to have to fork anything or look for other options if it can be avoided.

yeah me too. i have a project built using tsyringe that id rather not have to move to inversify.

**Comment by vajahath** [comment]
may be we should ping them on twitter I guess..

**Comment by AmauryD** [comment]
Just forked and published to npm while waiting for the real package to update. [@triptyk/tsyringe ](https://www.npmjs.com/package/@triptyk/tsyringe).

**Comment by jhechtf** [comment]
@Xapphire13 Hoping to get some traction on getting the changes for TS5 support published.

**Comment by wielski** [comment]
To avoid type error just place `.d.ts` file in root of your project, like `tsyringe.types.d.ts` with following contents:

```
import * as tsyringe from 'tsyringe';

// Overwrite wrong declaration from tryringe, support typescript 5.x
// Remove this when library will update
declare module 'tsyringe' {
  declare function inject(
    token: tsyringe.InjectionToken<any>
  ): (target: any, propertyKey: string | symbol | undefined, parameterIndex: number) => any;
}
```

It's a little hackish but at least it will work until package will be updated

**Comment by m-salman-afzal** [comment]
> To avoid type error just place `.d.ts` file in root of your project, like `tsyringe.types.d.ts` with following contents:
> 
> ```
> import * as tsyringe from 'tsyringe';
> 
> // Overwrite wrong declaration from tryringe, support typescript 5.x
> // Remove this when library will update
> declare module 'tsyringe' {
>   declare function inject(
>     token: tsyringe.InjectionToken<any>
>   ): (target: any, propertyKey: string | symbol | undefined, parameterIndex: number) => any;
> }
> ```
> 
> It's a little hackish but at least it will work until package will be updated

@wielski It works fine if you are just using tsc. Even VS code has no problems. But ts-node or ts-node-dev are giving the same error like it can't find our custom file. Any workaround that?

**Comment by m-salman-afzal** [comment]
got it working with --files flag that checks with the typeRoots option in tsconfig

**Comment by jhechtf** [comment]
Any comment about when this is going to be released? At this rate i might as well move myself to Inversify.

**Comment by vajahath** [comment]
I'm wondering what other DI frameworks are being used in other popular microsoft projects because this is not actively maintained. tsc itself is from ms. Many projects moved to tsc5 (eg vscode). So these tsc5 moved projects' di lib could be not tsyringe, something else. Not sure if we are missing something.

**Comment by woehrl01** [comment]
There was a new release published today to npm, in 4.8.0 this change is now included.

**Comment by MeltingMosaic** [comment]
Sorry folks - I thought I published a new version when I merged the change. Turns out, I did not.

**Comment by m-salman-afzal** [comment]
I think this issue can be closed now.

**Comment by Frikki** [comment]
@MeltingMosaic So, will the new version be published?

**Comment by stevenulibarri** [comment]
@Frikki
It was published; npm shows v4.8.0 now (published on 6/20) but here on github the releases page hasn't been updated since v4.4.0.

**Comment by JobaDiniz** [comment]
I'm using typescript 5 and tried to use this library, but VSCode complains with:

![image](https://github.com/microsoft/tsyringe/assets/165290/8c8928ac-03e4-4f3e-949f-3c3823ec93d5)
*Unable to resolve signature of class decorator when called as an expression*

![image](https://github.com/microsoft/tsyringe/assets/165290/913cc822-ad3d-4899-83e8-cd624b6c222b)
*Decorators are not valid here*

**tsconfig.json**
```json
"compilerOptions": {
      "experimentalDecorators": true,
      "emitDecoratorMetadata": true,
      "baseUrl": "src/",
      "declaration": true,
      "declarationDir": "build/",
      "emitDeclarationOnly": true,
      "esModuleInterop": true,
      "target": "ESNext",
      "useDefineForClassFields": true,
      "module": "ESNext",
      "lib": [
         "ESNext",
         "DOM",
         "DOM.Iterable",
         "ES2021.String"
      ],
      "moduleResolution": "Node",
      "strict": true,
      "resolveJsonModule": true,
      "isolatedModules": true,
      "noUnusedLocals": false,
      "noUnusedParameters": false,
      "noImplicitReturns": true,
      "skipLibCheck": true,
      "typeRoots": [
         "node_modules/@types"
      ]
```

Not sure how to move forward :(

**Comment by pclokcer** [comment]
> I'm using typescript 5 and tried to use this library, but VSCode complains with:
> 
> ![image](https://private-user-images.githubusercontent.com/165290/298171229-8c8928ac-03e4-4f3e-949f-3c3823ec93d5.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTY1OTgzNjMsIm5iZiI6MTcxNjU5ODA2MywicGF0aCI6Ii8xNjUyOTAvMjk4MTcxMjI5LThjODkyOGFjLTAzZTQtNGYzZS05NDlmLTNjMzgyM2VjOTNkNS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNTI1JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDUyNVQwMDQ3NDNaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1lMjYzZDU2NWRhNjhjMDU3ZGVhMGVjNzQ1OWUyMDc4OTU0ZTc0MmMxOTdmNDM0YTAxMGQ1MWY1MzRiMjcyNzE2JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.z8l0X-nJm6v2RMdjiu11esp5YJZEO2UQnTsMhkVGQa0) _Unable to resolve signature of class decorator when called as an expression_
> 
> ![image](https://private-user-images.githubusercontent.com/165290/298171324-913cc822-ad3d-4899-83e8-cd624b6c222b.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTY1OTgzNjMsIm5iZiI6MTcxNjU5ODA2MywicGF0aCI6Ii8xNjUyOTAvMjk4MTcxMzI0LTkxM2NjODIyLWFkM2QtNDg5OS04M2U4LWNkNjI0YjZjMjIyYi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNTI1JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDUyNVQwMDQ3NDNaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1lZDIyZDc2ZmYyZjMxNzkyZDZiYzM2ZDAwM2VhYmM3ODZjOWM0Yzc1ZjQ5N2E5ZTQ0YWQxZWI5ZjU3YWM5N2RlJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.Dwg0tRtcVJg2EftSh-NVudORlK8iplTvHgNzLPUeKZw) _Decorators are not valid here_
> 
> **tsconfig.json**
> 
> ```json
> "compilerOptions": {
>       "experimentalDecorators": true,
>       "emitDecoratorMetadata": true,
>       "baseUrl": "src/",
>       "declaration": true,
>       "declarationDir": "build/",
>       "emitDeclarationOnly": true,
>       "esModuleInterop": true,
>       "target": "ESNext",
>       "useDefineForClassFields": true,
>       "module": "ESNext",
>       "lib": [
>          "ESNext",
>          "DOM",
>          "DOM.Iterable",
>          "ES2021.String"
>       ],
>       "moduleResolution": "Node",
>       "strict": true,
>       "resolveJsonModule": true,
>       "isolatedModules": true,
>       "noUnusedLocals": false,
>       "noUnusedParameters": false,
>       "noImplicitReturns": true,
>       "skipLibCheck": true,
>       "typeRoots": [
>          "node_modules/@types"
>       ]
> ```
> 
> Not sure how to move forward :(

I am facing same error

**Comment by shawnallen85** [comment]
Any update on this working with TS5?

**Comment by exuanbo** [comment]
Hey everyone, I have been working on a library called [`di-wise`](https://github.com/exuanbo/di-wise). It has a similar but more powerful API that uses ECMAScript Stage 3 decorators introduced in TS 5. Welcome to try it out.
